File: libs/pipeline-expression/src/catalog.rs
//! Typed signatures for the canonical Pipeline Builder transforms +
//! the scalar helpers visible inside expression bodies.
//!
//! Two surfaces:
//!
//! * [`scalar_signature`] — looks up a scalar function by name (used
//!   inside [`crate::infer`] when the expression contains a call). It
//!   returns the arity, the argument types it accepts, and the result
//!   type.
//! * [`transform_signature`] — looks up the nine canonical pipeline
//!   transforms documented in
//!   `docs_original_palantir_foundry/foundry-docs/Data connectivity & integration/Workflows/Building pipelines/Getting started/Create a dataset batch pipeline with Pipeline Builder.md`
//!   plus the streaming/incremental docs (`window`, `pivot`).

use crate::types::PipelineType;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScalarSignature {
    pub name: &'static str,
    pub params: Vec<ParamConstraint>,
    pub result: ResultRule,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ParamConstraint {
    /// Argument must be exactly this type.
    Exactly(PipelineType),
    /// Argument must be promotable to this type.
    Promotable(PipelineType),
    /// Argument must satisfy a family predicate.
    Numeric,
    Textual,
    Temporal,
    /// Anything goes — used for `cast` and runtime-only checks.
    Any,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ResultRule {
    Fixed(PipelineType),
    /// Promote to the LUB of the listed argument indexes.
    PromoteOf(Vec<usize>),
    /// Result is the type literal extracted from arg N (used for
    /// `cast(value, type_literal)`).
    TypeFromStringArg(usize),
}

pub fn scalar_signature(name: &str) -> Option<ScalarSignature> {
    let n = name.to_ascii_lowercase();
    Some(match n.as_str() {
        // String helpers.
        "title_case" | "lower" | "upper" | "trim" | "clean_string" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Textual],
            result: ResultRule::Fixed(PipelineType::String),
        },
        "concat" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Textual, ParamConstraint::Textual],
            result: ResultRule::Fixed(PipelineType::String),
        },
        // Numeric helpers.
        "abs" | "floor" | "ceil" | "round" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Numeric],
            result: ResultRule::PromoteOf(vec![0]),
        },
        // Temporal.
        "to_date" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Textual],
            result: ResultRule::Fixed(PipelineType::Date),
        },
        "to_timestamp" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Textual],
            result: ResultRule::Fixed(PipelineType::Timestamp),
        },
        // Type cast: special-cased — second arg must be a STRING literal
        // naming the target type.
        "cast" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Any, ParamConstraint::Exactly(PipelineType::String)],
            result: ResultRule::TypeFromStringArg(1),
        },
        // Boolean.
        "is_null" | "is_not_null" => ScalarSignature {
            name: leak(&n),
            params: vec![ParamConstraint::Any],
            result: ResultRule::Fixed(PipelineType::Boolean),
        },
        // Geometry helpers.
        "geom_within" => ScalarSignature {
            name: leak(&n),
            params: vec![
                ParamConstraint::Exactly(PipelineType::Geometry),
                ParamConstraint::Exactly(PipelineType::Geometry),
            ],
            result: ResultRule::Fixed(PipelineType::Boolean),
        },
        _ => return None,
    })
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TransformSignature {
    pub name: &'static str,
    /// The transform either takes one input dataset or two (join/union).
    pub min_inputs: usize,
    pub max_inputs: usize,
    /// Required keys on the node `config` JSON object (informative).
    pub required_config_keys: &'static [&'static str],
}

pub fn transform_signature(name: &str) -> Option<TransformSignature> {
    let n = name.to_ascii_lowercase();
    Some(match n.as_str() {
        "cast" => TransformSignature {
            name: "cast",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["columns"],
        },
        "title_case" => TransformSignature {
            name: "title_case",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["columns"],
        },
        "clean_string" => TransformSignature {
            name: "clean_string",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["columns"],
        },
        "filter" => TransformSignature {
            name: "filter",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["predicate"],
        },
        "join" => TransformSignature {
            name: "join",
            min_inputs: 2,
            max_inputs: 2,
            required_config_keys: &["how", "on"],
        },
        "union" => TransformSignature {
            name: "union",
            min_inputs: 2,
            max_inputs: usize::MAX,
            required_config_keys: &[],
        },
        "group_by" => TransformSignature {
            name: "group_by",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["keys", "aggregations"],
        },
        "window" => TransformSignature {
            name: "window",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["partition_by", "order_by"],
        },
        "pivot" => TransformSignature {
            name: "pivot",
            min_inputs: 1,
            max_inputs: 1,
            required_config_keys: &["pivot_column", "value_column"],
        },
        _ => return None,
    })
}

/// Parse a string like `"INTEGER"` / `"ARRAY<INTEGER>"` into a
/// [`PipelineType`]. Used by `cast(value, "INTEGER")`. Returns `None`
/// when the literal isn't a recognised type name.
pub fn parse_type_literal(literal: &str) -> Option<PipelineType> {
    let trimmed = literal.trim();
    let upper = trimmed.to_ascii_uppercase();
    match upper.as_str() {
        "BOOLEAN" => Some(PipelineType::Boolean),
        "INTEGER" => Some(PipelineType::Integer),
        "LONG" => Some(PipelineType::Long),
        "DOUBLE" => Some(PipelineType::Double),
        "DECIMAL" => Some(PipelineType::Decimal),
        "STRING" => Some(PipelineType::String),
        "DATE" => Some(PipelineType::Date),
        "TIMESTAMP" => Some(PipelineType::Timestamp),
        "GEOMETRY" => Some(PipelineType::Geometry),
        _ => {
            if let Some(rest) = upper.strip_prefix("ARRAY<").and_then(|s| s.strip_suffix('>')) {
                return parse_type_literal(rest).map(PipelineType::array_of);
            }
            None
        }
    }
}

/// Static-string interner so `ScalarSignature::name` stays `&'static`.
fn leak(s: &str) -> &'static str {
    Box::leak(s.to_owned().into_boxed_str())
}

File: libs/pipeline-expression/catalog.go
package pipelineexpression

import (
	"math"
	"strings"
)

// ParamConstraintKind tags a [ParamConstraint].
type ParamConstraintKind int

const (
	// ParamExactly — argument must be exactly the listed type.
	ParamExactly ParamConstraintKind = iota
	// ParamPromotable — argument must be promotable to the listed type.
	ParamPromotable
	// ParamNumeric — argument must satisfy [PipelineType.IsNumeric].
	ParamNumeric
	// ParamTextual — argument must satisfy [PipelineType.IsTextual].
	ParamTextual
	// ParamTemporal — argument must satisfy [PipelineType.IsTemporal].
	ParamTemporal
	// ParamAny — anything goes (used for `cast` and runtime-only checks).
	ParamAny
)

// ParamConstraint describes the type constraint applied to a single
// scalar argument.
type ParamConstraint struct {
	Kind ParamConstraintKind
	Type PipelineType // populated for ParamExactly / ParamPromotable
}

// Equal compares two constraints structurally.
func (p ParamConstraint) Equal(other ParamConstraint) bool {
	if p.Kind != other.Kind {
		return false
	}
	if p.Kind == ParamExactly || p.Kind == ParamPromotable {
		return p.Type.Equal(other.Type)
	}
	return true
}

// ResultRuleKind tags a [ResultRule].
type ResultRuleKind int

const (
	// ResultFixed — result is the listed type.
	ResultFixed ResultRuleKind = iota
	// ResultPromoteOf — result is the LUB of the listed argument indexes.
	ResultPromoteOf
	// ResultTypeFromStringArg — result is the type literal extracted
	// from arg N (used for `cast(value, type_literal)`).
	ResultTypeFromStringArg
)

// ResultRule describes how a scalar's result type is computed.
type ResultRule struct {
	Kind  ResultRuleKind
	Type  PipelineType // populated for ResultFixed
	Args  []int        // populated for ResultPromoteOf
	Index int          // populated for ResultTypeFromStringArg
}

// ScalarSignature describes a scalar helper visible inside expression
// bodies.
type ScalarSignature struct {
	Name   string
	Params []ParamConstraint
	Result ResultRule
}

// ScalarSignatureFor looks up a scalar function by name. Returns
// (zero, false) when the name is not registered.
func ScalarSignatureFor(name string) (ScalarSignature, bool) {
	n := strings.ToLower(name)
	switch n {
	// String helpers.
	case "title_case", "lower", "upper", "trim", "clean_string":
		return ScalarSignature{
			Name:   n,
			Params: []ParamConstraint{{Kind: ParamTextual}},
			Result: ResultRule{Kind: ResultFixed, Type: StringType()},
		}, true
	case "concat":
		return ScalarSignature{
			Name:   n,
			Params: []ParamConstraint{{Kind: ParamTextual}, {Kind: ParamTextual}},
			Result: ResultRule{Kind: ResultFixed, Type: StringType()},
		}, true
	// Numeric helpers.
	case "abs", "floor", "ceil", "round":
		return ScalarSignature{
			Name:   n,
			Params: []ParamConstraint{{Kind: ParamNumeric}},
			Result: ResultRule{Kind: ResultPromoteOf, Args: []int{0}},
		}, true
	// Temporal.
	case "to_date":
		return ScalarSignature{
			Name:   n,
			Params: []ParamConstraint{{Kind: ParamTextual}},
			Result: ResultRule{Kind: ResultFixed, Type: DateType()},
		}, true
	case "to_timestamp":
		return ScalarSignature{
			Name:   n,
			Params: []ParamConstraint{{Kind: ParamTextual}},
			Result: ResultRule{Kind: ResultFixed, Type: TimestampType()},
		}, true
	// Type cast: special-cased — second arg must be a STRING literal
	// naming the target type.
	case "cast":
		return ScalarSignature{
			Name: n,
			Params: []ParamConstraint{
				{Kind: ParamAny},
				{Kind: ParamExactly, Type: StringType()},
			},
			Result: ResultRule{Kind: ResultTypeFromStringArg, Index: 1},
		}, true
	// Boolean.
	case "is_null", "is_not_null":
		return ScalarSignature{
			Name:   n,
			Params: []ParamConstraint{{Kind: ParamAny}},
			Result: ResultRule{Kind: ResultFixed, Type: BooleanType()},
		}, true
	// Geometry helpers.
	case "geom_within":
		return ScalarSignature{
			Name: n,
			Params: []ParamConstraint{
				{Kind: ParamExactly, Type: GeometryType()},
				{Kind: ParamExactly, Type: GeometryType()},
			},
			Result: ResultRule{Kind: ResultFixed, Type: BooleanType()},
		}, true
	}
	return ScalarSignature{}, false
}

// TransformSignature describes one of the canonical Pipeline Builder
// transforms.
type TransformSignature struct {
	Name string
	// MinInputs / MaxInputs bound the upstream-node fan-in.
	MinInputs int
	MaxInputs int
	// RequiredConfigKeys lists the keys the node `config` JSON object
	// must expose (informative).
	RequiredConfigKeys []string
}

// TransformMaxInputsUnbounded is the sentinel used for transforms that
// accept any number of upstream inputs (mirrors Rust `usize::MAX`).
const TransformMaxInputsUnbounded = math.MaxInt

// TransformSignatureFor looks up a transform by name. Returns
// (zero, false) when the name is not registered.
func TransformSignatureFor(name string) (TransformSignature, bool) {
	n := strings.ToLower(name)
	switch n {
	case "cast":
		return TransformSignature{
			Name: "cast", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"columns"},
		}, true
	case "title_case":
		return TransformSignature{
			Name: "title_case", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"columns"},
		}, true
	case "clean_string":
		return TransformSignature{
			Name: "clean_string", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"columns"},
		}, true
	case "filter":
		return TransformSignature{
			Name: "filter", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"predicate"},
		}, true
	case "join":
		return TransformSignature{
			Name: "join", MinInputs: 2, MaxInputs: 2,
			RequiredConfigKeys: []string{"how", "on"},
		}, true
	case "union":
		return TransformSignature{
			Name: "union", MinInputs: 2, MaxInputs: TransformMaxInputsUnbounded,
			RequiredConfigKeys: []string{},
		}, true
	case "group_by":
		return TransformSignature{
			Name: "group_by", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"keys", "aggregations"},
		}, true
	case "window":
		return TransformSignature{
			Name: "window", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"partition_by", "order_by"},
		}, true
	case "pivot":
		return TransformSignature{
			Name: "pivot", MinInputs: 1, MaxInputs: 1,
			RequiredConfigKeys: []string{"pivot_column", "value_column"},
		}, true
	}
	return TransformSignature{}, false
}

// ParseTypeLiteral parses a string like "INTEGER" / "ARRAY<INTEGER>"
// into a [PipelineType]. Used by `cast(value, "INTEGER")`. Returns
// (zero, false) when the literal isn't a recognised type name.
func ParseTypeLiteral(literal string) (PipelineType, bool) {
	trimmed := strings.TrimSpace(literal)
	upper := strings.ToUpper(trimmed)
	switch upper {
	case "BOOLEAN":
		return BooleanType(), true
	case "INTEGER":
		return IntegerType(), true
	case "LONG":
		return LongType(), true
	case "DOUBLE":
		return DoubleType(), true
	case "DECIMAL":
		return DecimalType(), true
	case "STRING":
		return StringType(), true
	case "DATE":
		return DateType(), true
	case "TIMESTAMP":
		return TimestampType(), true
	case "GEOMETRY":
		return GeometryType(), true
	}
	if strings.HasPrefix(upper, "ARRAY<") && strings.HasSuffix(upper, ">") {
		inner := upper[len("ARRAY<") : len(upper)-1]
		t, ok := ParseTypeLiteral(inner)
		if !ok {
			return PipelineType{}, false
		}
		return ArrayOf(t), true
	}
	return PipelineType{}, false
}
